test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -240,13 +248,23 @@ public InternalTestCluster(String nodeMode, long clusterSeed, Path baseDir,
 
         Random random = new Random(clusterSeed);
 
-        this.numSharedAllRolesNodes = RandomInts.randomIntBetween(random, minNumDataNodes, maxNumDataNodes);
-        assert this.numSharedAllRolesNodes >= 0;
+        this.numSharedDataNodes = RandomInts.randomIntBetween(random, minNumDataNodes, maxNumDataNodes);
```

Comment:
the subtle difference is that data nodes can be data only or data and master, while master ones are master only. I think renaming would make this clearer.